Introduce `nix derivation instantiate`

layus opened the pull request:
This is similar to `nix-instantiate` with support for installables (and flakes).

<!--

IMPORTANT

Nix is a non-trivial project, so for your contribution to be successful,
it really is important to follow the contributing guidelines:

https://github.com/NixOS/nix/blob/master/CONTRIBUTING.md

Even if you've contributed to open source before, take a moment to read it,
so you understand the process and the expectations.

- what information to include in commit messages
- proper attribution
- volunteering contributions effectively
- how to get help and our review process.

-->

# Motivation
<!-- Briefly explain what the change is about and why it is desirable. -->

The ability to obtain a derivation with the new cli has never been really implemented. https://github.com/NixOS/nix/issues/3908#issuecomment-1186633661 is the best work-around, but it does not add a gc root for the generated paths.

This brings a dedicated entry-point with the right semantics instead of the workaround above.

# Context
<!-- Provide context. Reference open issues if available. -->

# Fixes https://github.com/NixOS/nix/issues/3908

To some extent, this also addresses https://github.com/NixOS/nix/issues/7138 because outputs can easily be rooted. This brings that capability to store derivations.

<!-- Non-trivial change: Briefly outline the implementation strategy. -->

<!-- Invasive change: Discuss alternative designs or approaches you considered. -->

<!-- Large change: Provide instructions to reviewers how to read the diff. -->

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

Mic92 reviewed (changes requested):
Looks overall like a clean subset of `nix-instantiate`.
However some integration test would be nice and should be easy to add (with and without JSON).

Mic92 commented on src/nix/derivation-instantiate.cc:
```diff
@@ -0,0 +1,93 @@
+
+static nlohmann::json storePathSetToJSON(const StorePathSet & paths)
+{
+    auto res = nlohmann::json::array();
```
Maybe make this an object and put `paths` as a key in there. This makes it extensible in future.

fgaz reviewed (commented):
I think `instantiate` should not be placed under the `derivation` namespace. As I see it, `derivation` commands should operate on derivations and should not have to know about expressions/flakes/installables. This command however operates on installables to produce a derivation.

See also #7868 and #7903, cc @Ericson2314

layus commented:
@fgaz All right, but where then ?

layus commented:
The best place I can think of right now is `nix build --derivation`. But one may prefer `nix instantiate`. This is a trivial change, but how to get an agreement ? Should I organize a poll ?

fgaz commented:
> The best place I can think of right now is `nix build --derivation`. But one may prefer `nix instantiate`.

Both look fine to me. Another option is `nix eval --derivation` (`eval` instantiates the derivation from the expression anyway, it just isn't able to print it because `.drvPath` is an implementation detail).

> The best place I can think of right now is `nix build --derivation`. But one may prefer `nix instantiate`. This is a trivial change, but how to get an agreement ? Should I organize a poll ?

I think for now we can just wait for a nix maintainer to give their opinion.

edolstra commented:
I think `nix derivation instantiate` is preferable since we don't want to pollute the top-level namespace with a utility command like this. Also, maybe it should be `nix derivation create` to avoid the "instantiate" terminology from the old CLI. (Or `nix derivation add`, but that one is already taken. It should really have been called `nix derivation import`.)

BTW, the reason that the new CLI doesn't have an equivalent for `nix-instantiate` is because ideally we would get rid of store derivations in the future. The Nix client should just send an in-memory representation of derivations to the daemon, without the need to materialize them on disk. So the fewer places where we expose .drvs in the new CLI, the better.

roberth commented:
> So the fewer places where we expose .drvs in the new CLI, the better.

We're already doing a good job at this, but in this new scenario, we'll still need the functionality of writing `.drv`s somewhere not in memory, for analysis by other tools, such as `nix-diff`. While all other commands could skip the instantiation into the actual store, this command could be used to do it anyway.
Alternatively or in addition, this functionality could be added to all commands by means of a flag, such as `--instantiate` and/or `--instantiation-store`.
Before any of that, we have to be sure whether instantiation means the act of computing the derivations (hashing them, etc), which is quite independent of evaluating the derivation arguments, especially when we have hash placeholder string thunks, or whether it is the act of writing them out in a somewhat user facing way.
Or do we define instantiation as computing the derivation _of the `outPath`_  to further diminish the role of actual drv files? That is in contrast to `nix eval` which does not care about derivations and outputs any more than the language itself does.
(I assumed #6507; `s/outPath/drvPath/` otherwise.)

layus commented:
Should it be named `nix derivation add-to-store` or `nix derivation store` ? These would be more explicit that it is going to write the derivation to the store, right ?
I understand the desire to move away from .drv files, but these are a core part of many workflows and there is still no way to properly write one int the store and get a gc-root on it.

Should I add the feature to the old nix-instantiate command instead ?

roberth commented:
Hold on, shouldn't `nix derivation` commands be store-level only?
Not 100% sure we decided that. @Ericson2314 would know?

edolstra commented:
> Hold on, shouldn't nix derivation commands be store-level only?

No, I think commands should work on as many installable types as possible. We definitely want people to be able to say `nix derivation instantiate nixpkgs#hello` (just like they can currently do `nix derivation show nixpkgs#hello`).

tomberek commented:
Discussed in meeting:
- rename to `nix derivation instantiate`
-

layus commented:
@tomberek Thanks for making a decision. I updated the PR to reflect the naming choice.
It seems there may have been a second point, but you did not fill it. Did you discuss anything else ?

layus commented:
Oh, just rebased, that was probably the second point.

Ericson2314 commented:
`nix derivation` should be store-layer only. However, this could technically pass if the idea is that store-layer commands still get transparently "upgraded" in full Nix into talking "full" installables and not just "store layer installables". We have gone back-and-forth on that.

Mic92 commented:
> @tomberek Thanks for making a decision. I updated the PR to reflect the naming choice. It seems there may have been a second point, but you did not fill it. Did you discuss anything else ?

No. There was none.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-09-30-nix-team-meeting-minutes-182/53814/1

7c6f434c commented:
> We're already doing a good job at this, but in this new scenario, we'll still need the functionality of writing .drvs somewhere not in memory, for analysis by other tools, such as nix-diff.

Also, derivations provide some natural workflows for build-dependency GC-pinning.

edolstra commented on src/nix/derivation-instantiate.md:
```diff
@@ -0,0 +1,39 @@
+
+- `--json`
+
+  Dump a JSON object whose keys are the generated store derivations instread of
```
```suggestion
  Dump a JSON object whose keys are the generated store derivations instead of
```

roberth replied:
It would be preferable to return a list, so that multiple installables can be distinguished by their position, reflecting that of the command line.

This is also how `nix build` behaves.

Nice to have:
It would be possible to add an `outputs` field as well, with unbuilt paths. Only input-addressed output paths and fixed-output path are known ahead of time (not ca or impure), so they should be in `.drv^out` form or a JSON representation of that. Let's save that for a follow-up PR.

Thought (no action needed):
It also touches on https://github.com/NixOS/nix/issues/6507, but it's probably fine for this only work on derivations and not package attrsets (which aren't supported in the CLI yet).

layus replied:
The "defaultOutput" name would be very useful.
I think the output should contain roughly what we store in the eval cache database. These two things are similar.

edolstra commented on src/nix/derivation-instantiate.md:
```diff
@@ -0,0 +1,39 @@
+- `--json`
+
+  Dump a JSON object whose keys are the generated store derivations instread of
+  printing them directly on the output.
```
The JSON output format should be documented somewhere.

Mic92 replied:
Maybe some example would help.

edolstra commented on src/nix/derivation-instantiate.cc:
```diff
@@ -0,0 +1,93 @@
+    return res;
+}
+
+// TODO deduplicate with other code also setting such out links.
```
Would be nice to deduplicate this, but that can also be done in a future PR.

edolstra replied:
See [commit ](https://github.com/NixOS/nix/pull/11657/commits/7f6d006bebfb56b2967165c7f2e8c940abafac24) which factors out `createOutLinks()`.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-10-07-nix-team-meeting-minutes-184/54046/1

layus commented:
I have reworked output. It now uses a list instead of a set. Documentation and tests updated accordingly.

(also rebased, because I thought that may fix running tests in meson, but no. I still need to run `meson --reconfigure` to make `meson test` pick up changes to files in test/functional. Weird)

edolstra commented:
>  I still need to run meson --reconfigure to make meson test pick up changes to files in test/functional. Weird)

@Ericson2314 ^

Ericson2314 commented:
@layus 

This clashes with RFC 92 a bit, because it is unclear what it should do in the case of derivations that are themselves the build outputs of derivations.

We don’t want to do this now in light of those uncertainties, and would much prefer it if you could work on the plan outlined in https://github.com/NixOS/nix/issues/7261

We would also appreciate advice on how to make issues like this more prominant so contributors like you don’t end up spending a bunch of time implementing something that overlaps with an agreed-upon plan that’s just waiting to be implemented.

Sorry for the confusing discussion response from us on this, and the delayed disapproval.

layus commented:
@Ericson2314 
It's always better to have a clear no than uncertainties. That being said, uncertainties _and_ a no, that is a bit though.

Regarding the second point, https://github.com/NixOS/nix/issues/7261 is not a plan, it is an issue about the inconsistent semantics of --derivation. I do not see any design document, and no design document for the new nix cli either. As outlined by the uncertainty about basically everything in this PR. There is no clear authority chain for decisions, roadmaps are fragmented over several repos, issues and prs and seem to mostly reside unwritten in a few ppls heads, not necessarily aligning btw.

> I think instantiate should not be placed under the derivation namespace. As I see it, derivation commands should operate on derivations and should not have to know about expressions/flakes/installables.

> I think nix derivation instantiate is preferable since we don't want to pollute the top-level namespace with a utility command like this. Also, maybe it should be nix derivation create to avoid the "instantiate" terminology from the old CLI. 

> ideally we would get rid of store derivations in the future.

> Discussed in meeting: - rename to `nix derivation instantiate`

> Hold on, shouldn't nix derivation commands be store-level only?

> No, I think commands should work on as many installable types as possible. 

> `nix derivation` should be store-layer only. (where is this "store layer" defined, it was "store level" above ?)

> We don’t want to do this now in light of those uncertainties (who is "we" ?)

To be clear, I do not think it is an issue about making some pages more prominent. Starting with having the nix team agree on these topics, writing it somewhere, and referencing it whenever needed would help a lot. #7261 first appeared here in the 29th message, long after some prominent and active members already chimed in.

--- 

As for the decision, I think we are sacrificing usability today for the sake of an ideal future that may not come soon. #7261 is already two years old for example and I am literally using [this snippet](https://github.com/NixOS/nix/issues/7138#issuecomment-2346649899) in production. Why not just merge it as-is, and deprecate it when a perfect replacement becomes available ?

In all cases, the answer cannot be a no. It should ideally come with a path forward. After reading and clicking links all over, it seems that the desired UX is `nix build foo.drvPath` instead of `nix derivation instantiate foo`. Easy enough to implement in some cases, probably some nasty corner cases. At least that is what I get from this item in https://github.com/NixOS/nix/issues/7261:

> [ ] Make foo.drvPath an installable so we always have a nice way to select the derivation for high-level installables.

As much as I would like to help, I am not sure I want to get started down that rabbit hole. From a nix language perspective these are just strings with context. What is the implementation supposed to do with that context ? We will for sure end up with strange corner cases where the context is inconsistent with the string. What if the string is just hard-coded ? There is a big difference between a .drv generated as a side-effect of evaluation, and other files. The second case requires more careful validation of the input.

---

Anyway, @Ericson2314, would you mind having a proper entry in the next nix team meeting minutes about this PR ? It was already discussed in two previous meetings, and your comment seems to come out of nowhere in that context.

Ericson2314 commented:
In linking #7261 I also buried the lede: with the `^..` installable syntax proposed there
```
nix derivation instantiate foo
```
should be the same thing as
```
nix build foo^..
```
